src/metldata/builtin_transformations/infer_relation/data_transform.py

```diff
@@ -0,0 +1,121 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"Logic for transforming data."
+
+from schemapack.spec.datapack import DataPack
+from schemapack.spec.schemapack import ClassDefinition, ClassRelation, SchemaPack
+
+from metldata.builtin_transformations.common.custom_types import (
+    MutableClassResources,
+    MutableDatapack,
+    ResourceId,
+)
+from metldata.builtin_transformations.common.utils import data_to_dict
+from metldata.builtin_transformations.infer_relation.config import (
+    InferRelationConfig,
+)
+from metldata.builtin_transformations.infer_relation.resolve_path import (
+    resolve_path,
+)
+from metldata.transform.exceptions import EvitableTransformationError
+
+
+def get_class_resources(
+    *, data: MutableDatapack, class_name: str
+) -> MutableClassResources:
+    """Extract the resources of a given class from the dictionary."""
+    resources = data["resources"].get(class_name)
+    if not resources:
+        raise EvitableTransformationError()
+    return resources
+
+
+def infer_data_relation(
+    *, data: DataPack, model: SchemaPack, transformation_config: InferRelationConfig
+) -> DataPack:
+    """Adds inferred relations to the data as given in the configuration."""
+    modified_data = data_to_dict(data)
+
+    # name of the class to which the relation will be added
+    class_name = transformation_config.class_name
+
+    # name of the relation to be added
+    relation_name = transformation_config.relation_name
+
+    path = transformation_config.relation_path
+    referenced_class = path.target
+
+    # the target prefix refers to resources that will be modified by the transformation
+    target_resources = get_class_resources(data=modified_data, class_name=class_name)
+
+    for resource_id, resource in target_resources.items():
+        relation_target_ids = resolve_path(
+            data=data, source_resource_id=resource_id, path=path
+        )
+        type_corrected_relation_target_ids = modify_relation_target_ids_type(
+            model, class_name, relation_name, relation_target_ids
+        )
+        # add a new relation to that resource
+        resource.setdefault("relations", {})[relation_name] = {
+            "targetClass": referenced_class,
+            "targetResources": type_corrected_relation_target_ids,
+        }
+    return DataPack.model_validate(modified_data)
+
+
+def modify_relation_target_ids_type(
+    model: SchemaPack,
+    class_name: str,
+    relation_name: str,
+    relation_target_ids: frozenset[ResourceId],
+) -> frozenset[ResourceId] | ResourceId | None:
+    """Post transformation modification of the relation target IDs type based on
+    relation specification of the transformed model.
```

Comment:
Could be formulated in a bit less abstract way, i.e. what exactly is the modification here and under what condition, so something like `unpack resource ID from set if a string is expected by the model spec`